File: windows/SicDic.Core/Calculator.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace SicDic.Core;

public sealed record SicResult(int Platelets, int Inr, int Sofa, int Total, bool ThresholdMet)
{
    public string Vector() => $"{Platelets},{Inr},{Sofa},{Total},{ThresholdMet.ToString().ToLowerInvariant()}";
}
public sealed record DicResult(int Platelets, int Dimer, int Pt, int Fibrinogen, int Total, bool ThresholdMet,
    bool NegativePtDelta, decimal DeltaPt, decimal DimerValue, decimal DimerUln)
{
    public string Vector() => $"{Platelets},{Dimer},{Pt},{Fibrinogen},{Total},{ThresholdMet.ToString().ToLowerInvariant()},{NegativePtDelta.ToString().ToLowerInvariant()}";
    public string RatioDisplay() => (DimerValue / DimerUln).ToString("F6", CultureInfo.InvariantCulture);
}
public sealed record SofaResult(int Respiratory, int Cardiovascular, int Hepatic, int Renal, int Total)
{
    public string Vector() => $"{Respiratory},{Cardiovascular},{Hepatic},{Renal},{Total}";
}

public static class Calculator
{
    private static readonly Regex NumberPattern = new("\\A" + Rules.NumberPattern + "\\z", RegexOptions.CultureInvariant, TimeSpan.FromSeconds(1));
    public static decimal Number(string? raw)
    {
        if (raw is null || !NumberPattern.IsMatch(raw.Trim())) throw new ArgumentException("INVALID_NUMBER");
        return decimal.Parse(raw.Trim().Replace(',', '.'), NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture);
    }
    private static decimal Positive(string raw)
    {
        decimal value = Number(raw);
        if (value <= 0) throw new ArgumentException("POSITIVE_REQUIRED");
        return value;
    }
    private static bool Flag(string raw) => raw switch { "true" => true, "false" => false, _ => throw new ArgumentException("EXPLICIT_BOOLEAN_REQUIRED") };
    public static SicResult Sic(string platelets, string inr, string sofa)
    {
        decimal s = Number(sofa);
        if (s != decimal.Truncate(s) || s > Rules.SOFA_MAX) throw new ArgumentException("FOUR_COMPONENT_SOFA_INTEGER_REQUIRED");
        int p = Rules.sicPlatelets(Number(platelets), 1), i = Rules.sicInr(Positive(inr), 1), o = Rules.sicSofa(s, 1);
        int total = p + i + o;
        return new(p, i, o, total, total >= Rules.SIC_THRESHOLD && p + i > Rules.SIC_COAG_MIN);
    }
    public static DicResult Dic(string platelets, string dimer, string uln, string patientPt, string controlPt,
        string fibrinogen, string fibrinogenUnit, bool comparable)
    {
        if (!comparable) throw new ArgumentException("DIMER_NOT_COMPARABLE");
        decimal d = Number(dimer), u = Positive(uln), delta = Positive(patientPt) - Positive(controlPt);
        decimal f = fibrinogenUnit switch { "g/L" => Number(fibrinogen), "mg/dL" => Number(fibrinogen) / Rules.fibrinogenMgDlPerGL, _ => throw new ArgumentException("INVALID_UNIT") };
        int p = Rules.dicPlatelets(Number(platelets), 1), dd = Rules.dicDimer(d, u), pt = Rules.dicPt(delta, 1), fib = Rules.dicFibrinogen(f, 1);
        int total = p + dd + pt + fib;
        return new(p, dd, pt, fib, total, total >= Rules.DIC_THRESHOLD, delta < 0, delta, d, u);
    }
    public static SofaResult Sofa(string pf, string supported, string bilirubin, string creatinine, string labUnit,
        string urine24h, string map, string dopamine, string epinephrine, string norepinephrine, string dobutamine, string durationConfirmed)
    {
        if (!Flag(durationConfirmed)) throw new ArgumentException("SOFA_TIME_WINDOW_REQUIRED");
        decimal hs = 1, ks = 1;
        if (labUnit == "umol/L") { hs = Rules.bilirubinUmolPerMgDl; ks = Rules.creatinineUmolPerMgDl; }
        else if (labUnit != "mg/dL") throw new ArgumentException("INVALID_UNIT");
        decimal ratio = Number(pf);
        int r = Flag(supported) ? Rules.respiratorySupported(ratio, 1) : Rules.respiratoryUnsupported(ratio, 1);
        int c = new[] { Rules.map(Positive(map), 1), Rules.dopamine(Number(dopamine), 1), Rules.epinephrine(Number(epinephrine), 1), Rules.epinephrine(Number(norepinephrine), 1), Rules.dobutamine(Number(dobutamine), 1) }.Max();
        int h = Rules.hepaticMgDl(Number(bilirubin), hs), k = Math.Max(Rules.renalMgDl(Number(creatinine), ks), Rules.urine24h(Number(urine24h), 1));
        return new(r, c, h, k, r + c + h + k);
    }
    public static string Convert(string raw, string quantity)
    {
        decimal divisor = quantity switch { "fibrinogen" => Rules.fibrinogenMgDlPerGL, "bilirubin" => Rules.bilirubinUmolPerMgDl, "creatinine" => Rules.creatinineUmolPerMgDl, _ => throw new ArgumentException("INVALID_QUANTITY") };
        return decimal.Round(Number(raw) / divisor, 6, MidpointRounding.AwayFromZero).ToString("0.######", CultureInfo.InvariantCulture);
    }
    public static string EvaluateVector(string op, string[] a)
    {
        try
        {
            return op switch
            {
                "sic" => Sic(a[0], a[1], a[2]).Vector(),
                "dic" => Dic(a[0], a[1], a[2], a[3], a[4], a[5], a[6], Flag(a[7])).Vector(),
                "sofa" => Sofa(a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8], a[9], a[10], a[11]).Vector(),
                "convert" => Convert(a[0], a[1]),
                _ => throw new ArgumentException("UNKNOWN_OPERATION")
            };
        }
        catch (ArgumentException) { return "ERROR"; }
    }
}

File: android/core/src/main/java/org/sicdic/core/Calculator.java
package org.sicdic.core;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.regex.Pattern;

/** Pure, offline clinical arithmetic. Numeric threshold flags are not diagnoses. */
public final class Calculator {
    private static final BigDecimal ONE = BigDecimal.ONE;
    private static final Pattern NUMBER = Pattern.compile(Rules.NUMBER_PATTERN);
    private Calculator() {}

    public static BigDecimal number(String raw) {
        if (raw == null || !NUMBER.matcher(raw.trim()).matches()) throw new IllegalArgumentException("INVALID_NUMBER");
        return new BigDecimal(raw.trim().replace(',', '.'));
    }
    private static BigDecimal positive(String raw) {
        BigDecimal value = number(raw);
        if (value.signum() <= 0) throw new IllegalArgumentException("POSITIVE_REQUIRED");
        return value;
    }
    private static boolean flag(String raw) {
        if (!"true".equals(raw) && !"false".equals(raw)) throw new IllegalArgumentException("EXPLICIT_BOOLEAN_REQUIRED");
        return Boolean.parseBoolean(raw);
    }
    public record Sic(int platelets, int inr, int sofa, int total, boolean thresholdMet) {
        public String vector() { return platelets+","+inr+","+sofa+","+total+","+thresholdMet; }
    }
    public record Dic(int platelets, int dimer, int pt, int fibrinogen, int total,
                      boolean thresholdMet, boolean negativePtDelta, BigDecimal deltaPt,
                      BigDecimal dimerValue, BigDecimal dimerUln) {
        public String vector() { return platelets+","+dimer+","+pt+","+fibrinogen+","+total+","+thresholdMet+","+negativePtDelta; }
        /** Display only; classification uses exact cross-multiplication. */
        public String ratioDisplay() { return dimerValue.divide(dimerUln, 6, RoundingMode.HALF_UP).toPlainString(); }
    }
    public record Sofa(int respiratory, int cardiovascular, int hepatic, int renal, int total) {
        public String vector() { return respiratory+","+cardiovascular+","+hepatic+","+renal+","+total; }
    }
    public static Sic sic(String platelets, String inr, String sofa) {
        BigDecimal s = number(sofa);
        if (s.stripTrailingZeros().scale() > 0 || s.compareTo(BigDecimal.valueOf(Rules.SOFA_MAX)) > 0)
            throw new IllegalArgumentException("FOUR_COMPONENT_SOFA_INTEGER_REQUIRED");
        int p = Rules.sicPlatelets(number(platelets), ONE);
        int i = Rules.sicInr(positive(inr), ONE);
        int o = Rules.sicSofa(s, ONE);
        int total = p+i+o;
        return new Sic(p,i,o,total,total >= Rules.SIC_THRESHOLD && p+i > Rules.SIC_COAG_MIN);
    }
    public static Dic dic(String platelets, String dimer, String uln, String patientPt, String controlPt,
                          String fibrinogen, String fibrinogenUnit, boolean comparable) {
        if (!comparable) throw new IllegalArgumentException("DIMER_NOT_COMPARABLE");
        BigDecimal d = number(dimer), u = positive(uln);
        BigDecimal delta = positive(patientPt).subtract(positive(controlPt));
        BigDecimal f = number(fibrinogen);
        if ("mg/dL".equals(fibrinogenUnit)) f = f.divide(Rules.fibrinogenMgDlPerGL);
        else if (!"g/L".equals(fibrinogenUnit)) throw new IllegalArgumentException("INVALID_UNIT");
        int p = Rules.dicPlatelets(number(platelets), ONE), dd = Rules.dicDimer(d,u);
        int pt = Rules.dicPt(delta,ONE), fib = Rules.dicFibrinogen(f,ONE);
        int total = p+dd+pt+fib;
        return new Dic(p,dd,pt,fib,total,total >= Rules.DIC_THRESHOLD,delta.signum()<0,delta,d,u);
    }
    /** Development helper policy is explicit in clinical-spec/RULES.md; no SOFA-2 substitution. */
    public static Sofa sofa(String pf, String supported, String bilirubin, String creatinine, String labUnit,
                            String urine24h, String map, String dopamine, String epinephrine,
                            String norepinephrine, String dobutamine, String durationConfirmed) {
        if (!flag(durationConfirmed)) throw new IllegalArgumentException("SOFA_TIME_WINDOW_REQUIRED");
        BigDecimal hepaticScale=ONE, renalScale=ONE;
        if ("umol/L".equals(labUnit)) { hepaticScale=Rules.bilirubinUmolPerMgDl; renalScale=Rules.creatinineUmolPerMgDl; }
        else if (!"mg/dL".equals(labUnit)) throw new IllegalArgumentException("INVALID_UNIT");
        BigDecimal ratio=number(pf);
        int r=flag(supported) ? Rules.respiratorySupported(ratio,ONE) : Rules.respiratoryUnsupported(ratio,ONE);
        int c=Rules.map(positive(map),ONE);
        c=Math.max(c,Rules.dopamine(number(dopamine),ONE));
        c=Math.max(c,Rules.epinephrine(number(epinephrine),ONE));
        c=Math.max(c,Rules.epinephrine(number(norepinephrine),ONE));
        c=Math.max(c,Rules.dobutamine(number(dobutamine),ONE));
        int h=Rules.hepaticMgDl(number(bilirubin),hepaticScale);
        int k=Math.max(Rules.renalMgDl(number(creatinine),renalScale),Rules.urine24h(number(urine24h),ONE));
        return new Sofa(r,c,h,k,r+c+h+k);
    }
    public static String convert(String raw, String quantity) {
        BigDecimal divisor = switch(quantity) {
            case "fibrinogen" -> Rules.fibrinogenMgDlPerGL;
            case "bilirubin" -> Rules.bilirubinUmolPerMgDl;
            case "creatinine" -> Rules.creatinineUmolPerMgDl;
            default -> throw new IllegalArgumentException("INVALID_QUANTITY");
        };
        return number(raw).divide(divisor, 6, RoundingMode.HALF_UP).stripTrailingZeros().toPlainString();
    }
    public static String evaluateVector(String op, String[] a) {
        try {
            return switch(op) {
                case "sic" -> sic(a[0],a[1],a[2]).vector();
                case "dic" -> dic(a[0],a[1],a[2],a[3],a[4],a[5],a[6],flag(a[7])).vector();
                case "sofa" -> sofa(a[0],a[1],a[2],a[3],a[4],a[5],a[6],a[7],a[8],a[9],a[10],a[11]).vector();
                case "convert" -> convert(a[0],a[1]);
                default -> throw new IllegalArgumentException("UNKNOWN_OPERATION");
            };
        } catch (IllegalArgumentException e) { return "ERROR"; }
    }
}
